src/main/java/io/vertx/core/json/pointer/JsonPointer.java

```diff
@@ -0,0 +1,236 @@
+   */
+  default @Nullable Object queryJsonOrDefault(Object object, Object defaultValue) {return queryOrDefault(object, JsonPointerIterator.JSON_ITERATOR, defaultValue); }
+
+  /**
+   * Query the provided object tracing each element walked during the query, including the first and the result (if any).<br/>
+   * The first element of the stream is objectToQuery and the last is the result, or the element before the first null was encountered
+   *
+   * @param objectToQuery the object to query
+   * @param iterator the json pointer iterator that provides the logic to access to the objectToQuery
+   * @return the stream of walked elements
+   */
+  @GenIgnore
+  Stream<Object> tracedQuery(Object objectToQuery, JsonPointerIterator iterator);
+
+  /**
+   * Write a value with the selected pointer. The path token "-" is handled as append to end of array <br/>
+   * If you need to write in Vert.x json data structures, use {@link JsonPointer#writeJson(Object, Object)} (Object)}<br/>
+   *
+   * @param objectToWrite object to write
+   * @param iterator the json pointer iterator that provides the logic to access to the objectToMutate
+   * @param newElement  object to insert
+   * @param createOnMissing create objects when missing a object key or an array index
+   * @return a reference to objectToWrite if the write was completed, a reference to newElement if the pointer is a root pointer, null if the write failed
+   */
+  Object write(Object objectToWrite, JsonPointerIterator iterator, Object newElement, boolean createOnMissing);
+
+  /**
+   * Write a value in the selected pointer. The path token "-" is handled as append to end of array. <br/>
+   * This function does not support root pointers.
+   *
+   * @param json json to query and write
+   * @param newElement json to insert
+   * @return a reference to json if the write was completed, a reference to newElement if the pointer is a root pointer, null if the write failed
+   */
+  default Object writeJson(Object json, Object newElement) { return writeJson(json, newElement, false); }
+
+  /**
+   * Write a value in the selected pointer. The path token "-" is handled as append to end of array. <br/>
+   * This function does not support root pointers.
+   *
+   * @param json json to query and write
+   * @param newElement json to insert
+   * @param createOnMissing create JsonObject when missing a object key or an array index
+   * @return a reference to json if the write was completed, a reference to newElement if the pointer is a root pointer, null if the write failed
+   */
+  default Object writeJson(Object json, Object newElement, boolean createOnMissing) {
+    return write(json, JsonPointerIterator.JSON_ITERATOR, newElement, createOnMissing);
+  }
+
+  /**
+   * Copy a JsonPointer
+   *
+   * @return
+   */
+  JsonPointer copy();
+
+  /**
+   * Build an empty JsonPointer
+   *
+   * @return a new empty JsonPointer
+   */
+  static JsonPointer create() {
+    return new JsonPointerImpl();
```

Comment:
why is it useful ?

Reply:
It's useful to create a new empty pointer

Reply:
why are empty pointer useful ? I thought they are immutable and created from a string parsed

Reply:
No, JsonPointer is mutable and offers the `append` methods to build the pointer programmatically without caring about encoding. I would avoid making it immutable because it's a waste of memory, in particular in some phases of json validator

Reply:
If we want to make JsonPointer immutable we can create a JsonPointerBuilder, but it seems a little bit overkilling to me. wdyt?

Reply:
The intent of having a separate `JsonPointer` class (rather than building a json pointer query API directly into `JsonObject` / `JsonArray`, with the query string re-interpreted for each new lookup) seems to be so that a `JsonPointer` is created once and reused wherever possible. Maybe then the immutable model makes more sense, since the creation costs can in many cases be amortized over multiple uses?

One thing I would like to see is the ability to pre-build a query with named parameter slots, and then be able to pass a `Map` of concrete arguments into the `query` functions. When a named parameter slot is reached while following the query path, the value of the parameter will be looked up in the map. If this query form were supported, it would dramatically reduce the number of `JsonPointer` objects you would need to create, since the vast majority of queries could be built just once on startup, rather than having to be re-build for each new query. This would allow `JsonPointer` to act sort of like a prepared statement in SQL.

Reply:
That said, for many uses, users will probably simply prefer never having to deal with `JsonPointer` directly, but rather just be able to do something like `jsonObject.query("/users/" + userId.replace("~", "~0").replace("/", "~1") + "/age")`, or maybe something like the following could be supported that supports automatic escaping: `jsonObject.query("/users", userId, "age")`. (It's a real shame the json pointer RFC escaping syntax is so weird...)

Except for the most performance-critical applications, direct querying of this form will simpler and more natural than having to work through a separate `JsonPointer` class.

Reply:
`JsonPointer` needs to exist because it can be used for querying also non Vert.x data structures (look at `JsonPointerIterator`) and the notation `JsonObject.query()` would not work because of polyglot (json objects in other languages supported by Vert.x are not instances of Vert.x `JsonObject`)

I don't like the argument of making immutable because we need some kind of builder (that adds an extra step to usage of the api) and it's not consistent with Vert.x Json APIs, where both JsonArray and JsonObject are mutable.

For the parameters, seems like a cool idea but I would like to keep it out of the scope of this PR. Since Json pointers are very simple to create, you can just create methods to generate the pointers you need, like:
```
public void createAgePointer(String userId) {
  return JsonPointer.create().append("users").append(userId).append("age")
}
```
And there is another problem: "{userId}" is a valid object key for json pointers, so that means that we should find some weird escaping too for it or we can allow it only using `appendParameter()`, making it harder to use/understand for people

Reply:
@slinkydeveloper All good points. You're right about escaping, so I guess `appendParameter()` is the only option. I still think it would be a good idea to be able to parameterize these. Users that don't want the complexity can ignore that part of the API.

I just submitted a PR, #2931, with the simple `query()` syntax I was suggesting. I didn't know that other languages don't use `JsonObject` / `JsonArray`. However, for JVM languages that have those classes, this makes the API of those classes much simpler to use for deep object graphs, so I still think it's worth it.